### PR TITLE
[Accounts for Children] Creation of Banner

### PR DESCRIPTION
**Pull request by anmode**
# Descripción

Created the banner to display in the u13 dashboard to remind them about to verify the accounts by their parents
This will be merged when backend PRs will be merged

# Checklist:

- [x] El código sigue la [guía de
      estilo](https://github.com/omegaup/omegaup/wiki/Coding-guidelines) de
      omegaUp.
- [x] Se corrieron todas las pruebas y pasaron.
- [x] Si se está agregando funcionalidad nueva, se agregaron pruebas.
- [x] Si el cambio es grande (> 200 líneas), hay que intentar partirlo en
      varios pull requests. De preferencia uno para los controladores + phpunit
      y luego otro para la interfaz.

### CODE REVIEW

File: frontend/server/src/Controllers/Session.php [hunk 1]
```diff
@@ -25,7 +25,7 @@ public function __construct() {
  * @psalm-type AssociatedIdentity=array{default: bool, username: string}
  * @psalm-type IdentityExt=array{classname: string, country_id: null|string, current_identity_school_id: int|null, gender: null|string, identity_id: int, language_id: int|null, name: null|string, password: null|string, state_id: null|string, user_id: int|null, username: string}
  * @psalm-type AuthIdentityExt=array{currentIdentity: IdentityExt, loginIdentity: IdentityExt}
- * @psalm-type CurrentSession=array{apiTokenId: int|null, associated_identities: list<AssociatedIdentity>, auth_token: null|string, cacheKey: null|string, classname: string, email: null|string, identity: \OmegaUp\DAO\VO\Identities|null, is_admin: bool, loginIdentity: \OmegaUp\DAO\VO\Identities|null, user: \OmegaUp\DAO\VO\Users|null, valid: bool}
+ * @psalm-type CurrentSession=array{apiTokenId: int|null, associated_identities: list<AssociatedIdentity>, auth_token: null|string, cacheKey: null|string, classname: string, email: null|string, identity: \OmegaUp\DAO\VO\Identities|null, is_admin: bool, loginIdentity: \OmegaUp\DAO\VO\Identities|null, user: \OmegaUp\DAO\VO\Users|null, valid: bool,pendingDaysForVerification: int|null}
```
Comment (pabo99) [inline frontend/server/src/Controllers/Session.php:28]: nit:
```suggestion
 * @psalm-type CurrentSession=array{apiTokenId: int|null, associated_identities: list<AssociatedIdentity>, auth_token: null|string, cacheKey: null|string, classname: string, email: null|string, identity: \OmegaUp\DAO\VO\Identities|null, is_admin: bool, loginIdentity: \OmegaUp\DAO\VO\Identities|null, user: \OmegaUp\DAO\VO\Users|null, valid: bool, pendingDaysForVerification: int|null}
```

File: frontend/server/src/Controllers/Session.php [hunk 2]
```diff
@@ -324,6 +325,19 @@ private static function getCurrentSessionImpl(
             $currentUser = \OmegaUp\DAO\Users::getByPK(
                 $currentIdentity->user_id
             );
+            $pendingDaysForVerification = null;
```
Comment (pabo99) [inline frontend/server/src/Controllers/Session.php:328]: This line should be placed before any condition, maybe in L305 it's ok

File: frontend/server/src/Controllers/Session.php [hunk 3]
```diff
@@ -324,6 +325,19 @@ private static function getCurrentSessionImpl(
...
+                    '-13 year',
+                    $now
+                ) && !$currentUser->parent_verified
+            ) {
+                $diff = $currentUser->parent_email_verification_deadline->time - $now;
+                $pendingDaysForVerification = floor($diff / (60 * 60 * 24));
+            }
             if (is_null($currentUser)) {
                 throw new \OmegaUp\Exceptions\NotFoundException('userNotExist');
             }
```
Comment (pabo99) [inline frontend/server/src/Controllers/Session.php:343]: This condition should be placed before the birth date validations

File: frontend/server/src/Controllers/Session.php [hunk 4]
```diff
@@ -324,6 +325,19 @@ private static function getCurrentSessionImpl(
...
+            $pendingDaysForVerification = null;
+            $now = \OmegaUp\Time::get();
+            if (
+                strtotime(
+                    $currentUser->birth_date
+                ) >= strtotime(
+                    '-13 year',
+                    $now
+                ) && !$currentUser->parent_verified
+            ) {
```
Comment (pabo99) [inline frontend/server/src/Controllers/Session.php:337]: Here we need validate that the birthdate is not null:
```suggestion
            if (
                !is_null(
                    $currentUser->birth_date
                ) && strtotime(
                    $currentUser->birth_date
                ) >= strtotime(
                    '-13 year',
                    $now
                ) && !$currentUser->parent_verified
            ) {
```

File: frontend/server/src/Controllers/Session.php [hunk 5]
```diff
@@ -324,6 +325,19 @@ private static function getCurrentSessionImpl(
...
+                $diff = $currentUser->parent_email_verification_deadline->time - $now;
+                $pendingDaysForVerification = floor($diff / (60 * 60 * 24));
```
Comment (pabo99) [inline frontend/server/src/Controllers/Session.php:339]: nit:
```suggestion
                $pendingDaysForVerification = intval(floor($diff / (60 * 60 * 24)));
```

File: frontend/server/src/Controllers/User.php [hunk 6]
```diff
@@ -10,7 +10,7 @@
...
  * @psalm-type AssociatedIdentity=array{username: string, default: bool}
- * @psalm-type CommonPayload=array{associatedIdentities: list<AssociatedIdentity>, currentEmail: string, currentName: null|string, currentUsername: string, gravatarURL128: string, gravatarURL51: string, isAdmin: bool, inContest: bool, isLoggedIn: bool, isMainUserIdentity: bool, isReviewer: bool, lockDownImage: string, navbarSection: string, omegaUpLockDown: bool, profileProgress: float, userClassname: string, userCountry: string, userTypes: list<string>}
+ * @psalm-type CommonPayload=array{associatedIdentities: list<AssociatedIdentity>, currentEmail: string, currentName: null|string, currentUsername: string, gravatarURL128: string, gravatarURL51: string, isAdmin: bool, inContest: bool, isLoggedIn: bool, isMainUserIdentity: bool, isReviewer: bool, lockDownImage: string, navbarSection: string, omegaUpLockDown: bool, profileProgress: float, userClassname: string, userCountry: string, userTypes: list<string>,pendingDaysForVerification: int|null}
```
Comment (pabo99) [inline frontend/server/src/Controllers/User.php:13]: nit:
```suggestion
 * @psalm-type CommonPayload=array{associatedIdentities: list<AssociatedIdentity>, currentEmail: string, currentName: null|string, currentUsername: string, gravatarURL128: string, gravatarURL51: string, isAdmin: bool, inContest: bool, isLoggedIn: bool, isMainUserIdentity: bool, isReviewer: bool, lockDownImage: string, navbarSection: string, omegaUpLockDown: bool, profileProgress: float, userClassname: string, userCountry: string, userTypes: list<string>, pendingDaysForVerification: int|null}
```

File: frontend/server/src/UITools.php [hunk 7]
```diff
@@ -4,8 +4,8 @@
...
- * @psalm-type CurrentSession=array{associated_identities: list<AssociatedIdentity>, valid: bool, email: string|null, user: \OmegaUp\DAO\VO\Users|null, identity: \OmegaUp\DAO\VO\Identities|null, classname: string, auth_token: string|null, is_admin: bool}
+ * @psalm-type CommonPayload=array{associatedIdentities: list<AssociatedIdentity>, currentEmail: string, currentName: null|string, currentUsername: string, gravatarURL128: string, gravatarURL51: string, isAdmin: bool, inContest: bool, isLoggedIn: bool, isMainUserIdentity: bool, isReviewer: bool, lockDownImage: string, navbarSection: string, omegaUpLockDown: bool, profileProgress: float, userClassname: string, userCountry: string, userTypes: list<string>,pendingDaysForVerification: int|null}
```
Comment (pabo99) [inline frontend/server/src/UITools.php:7]: nit:
```suggestion
 * @psalm-type CommonPayload=array{associatedIdentities: list<AssociatedIdentity>, currentEmail: string, currentName: null|string, currentUsername: string, gravatarURL128: string, gravatarURL51: string, isAdmin: bool, inContest: bool, isLoggedIn: bool, isMainUserIdentity: bool, isReviewer: bool, lockDownImage: string, navbarSection: string, omegaUpLockDown: bool, profileProgress: float, userClassname: string, userCountry: string, userTypes: list<string>, pendingDaysForVerification: int|null}
```

File: frontend/server/src/UITools.php [hunk 8]
```diff
@@ -4,8 +4,8 @@
...
+ * @psalm-type CurrentSession=array{associated_identities: list<AssociatedIdentity>, valid: bool, email: string|null, user: \OmegaUp\DAO\VO\Users|null, identity: \OmegaUp\DAO\VO\Identities|null, classname: string, auth_token: string|null, is_admin: bool,pendingDaysForVerification: int|null}
```
Comment (pabo99) [inline frontend/server/src/UITools.php:8]: nit:
```suggestion
 * @psalm-type CurrentSession=array{associated_identities: list<AssociatedIdentity>, valid: bool, email: string|null, user: \OmegaUp\DAO\VO\Users|null, identity: \OmegaUp\DAO\VO\Identities|null, classname: string, auth_token: string|null, is_admin: bool, pendingDaysForVerification: int|null}
```

File: frontend/templates/en.lang [hunk 9]
```diff
@@ -4,6 +4,7 @@ accountDeletionConfirmEmailBody = "User %(username) confirmed to delete their ac
 accountDeletionConfirmEmailSubject = "omegaup account deleted"
 accountDeletionRequestEmailBody = "User %(username) is requesting delete their account."
 accountDeletionRequestEmailSubject = "omegaup account deletion request"
+accountVerifyWarning = "Your parent or Guardian must approve your acount otherwise it will be deleted in"
```
Comment (pabo99) [inline frontend/templates/en.lang:7]: Here you can use string interpolation:
```suggestion
accountVerifyWarning = "Your parent or Guardian must approve your acount otherwise it will be deleted in %(days) days"
```
the same in the files es.lang and pt.lang

File: frontend/www/js/omegaup/common/index.ts [hunk 10]
```diff
@@ -2,9 +2,27 @@ import Homepage from '../components/homepage/Homepage.vue';
 import { OmegaUp } from '../omegaup';
 import { types } from '../api_types';
 import Vue from 'vue';
+import * as ui from '../ui';
+import T from '../lang';
 
 OmegaUp.on('ready', () => {
   const payload = types.payloadParsers.IndexPayload();
+  const commonPayload = types.payloadParsers.CommonPayload();
+  console.log(commonPayload.pendingDaysForVerification);
```
Comment (pabo99) [inline frontend/www/js/omegaup/common/index.ts:11]: We don't recommend print anything in client logs
```suggestion
```

File: frontend/www/js/omegaup/common/index.ts [hunk 11]
```diff
@@ -2,9 +2,27 @@ import Homepage from '../components/homepage/Homepage.vue';
...
+      T.accountVerifyWarning +
+        commonPayload.pendingDaysForVerification +
+        T.days,
+    );
+  } else if (commonPayload.pendingDaysForVerification <= 7) {
+    ui.warning(
+      T.accountVerifyWarning +
+        commonPayload.pendingDaysForVerification +
+        T.days,
+    );
```
Comment (pabo99) [inline frontend/www/js/omegaup/common/index.ts:24]: Here you can use string interpolation as well:
```suggestion
      ui.warning(
        ui.formatString(T.accountVerifyWarning, {
          days: commonPayload.pendingDaysForVerification,
        }),
      );
```

File: frontend/www/js/omegaup/common/index.ts [hunk 12]
```diff
@@ -2,9 +2,27 @@ import Homepage from '../components/homepage/Homepage.vue';
...
   const payload = types.payloadParsers.IndexPayload();
+  const commonPayload = types.payloadParsers.CommonPayload();
+  console.log(commonPayload.pendingDaysForVerification);
+
+  if (commonPayload.pendingDaysForVerification < 5) {
+    ui.warning(
+      T.accountVerifyWarning +
+        commonPayload.pendingDaysForVerification +
+        T.days,
+    );
```
Comment (pabo99) [inline frontend/www/js/omegaup/common/index.ts:18]: Here you can use string interpolation:
```suggestion
      ui.error(
        ui.formatString(T.accountVerifyWarning, {
          days: commonPayload.pendingDaysForVerification,
        }),
      );
```

File: frontend/www/js/omegaup/common/index.ts [hunk 13]
```diff
@@ -2,9 +2,27 @@ import Homepage from '../components/homepage/Homepage.vue';
...
 import Vue from 'vue';
+import * as ui from '../ui';
+import T from '../lang';
 
 OmegaUp.on('ready', () => {
   const payload = types.payloadParsers.IndexPayload();
+  const commonPayload = types.payloadParsers.CommonPayload();
+  console.log(commonPayload.pendingDaysForVerification);
+
+  if (commonPayload.pendingDaysForVerification < 5) {
```
Comment (pabo99) [inline frontend/www/js/omegaup/common/index.ts:13]: `pendingDaysForVerification` could be null, so, you need add a new condition before:
```js
if (commonPayload.pendingDaysForVerification != null) {
```

File: frontend/server/src/Controllers/Session.php [hunk 14]
```diff
@@ -327,6 +329,31 @@ private static function getCurrentSessionImpl(
...
+                                {
+                                $diff = $currentUser->parent_email_verification_deadline->time - $now;
+                                $pendingDaysForVerification = intval(
+                                    floor(
+                                        $diff / (60 * 60 * 24)
+                                    )
+                                );
+                                }
+                }
+            }
```
Comment (pabo99) [inline frontend/server/src/Controllers/Session.php:355]: `parent_email_verification_deadline` should be required here, because otherwise the banner never will be shown. It would be better throw an exception indicating that this parameter is missing
```suggestion
                if (
                    is_null($currentUser->parent_email_verification_deadline)
                ) {
                    throw new \OmegaUp\Exceptions\InvalidParameterException(
                        'parameterNotFound',
                        'parent_email_verification_deadline'
                    );     
                }
                $diff = $currentUser->parent_email_verification_deadline->time - $now;
                $pendingDaysForVerification = intval(
                    floor(
                        $diff / (60 * 60 * 24)
                    )
                );
```
.

Another validation that  should  be included is to block the account when `parent_email_verification_deadline` has expired. Maybe this is the right place for it.

File: frontend/templates/en.lang [hunk 15]
```diff
@@ -4,6 +4,7 @@ accountDeletionConfirmEmailBody = "User %(username) confirmed to delete their ac
 accountDeletionConfirmEmailSubject = "omegaup account deleted"
 accountDeletionRequestEmailBody = "User %(username) is requesting delete their account."
 accountDeletionRequestEmailSubject = "omegaup account deletion request"
+accountVerifyWarning = "Your parent or Guardian must approve your acount otherwise it will be deleted in  %(days) days"
```
Comment (pabo99) [inline frontend/templates/en.lang:7]: nit:
```suggestion
accountVerifyWarning = "Your parent or Guardian must approve your account otherwise it will be deleted in  %(days) days"
```

File: frontend/www/js/omegaup/common/index.ts [hunk 16]
```diff
@@ -2,9 +2,28 @@ import Homepage from '../components/homepage/Homepage.vue';
...
+import * as ui from '../ui';
+import T from '../lang';
 
 OmegaUp.on('ready', () => {
   const payload = types.payloadParsers.IndexPayload();
+  const commonPayload = types.payloadParsers.CommonPayload();
+
+  if (commonPayload.pendingDaysForVerification != null) {
+    if (commonPayload.pendingDaysForVerification < 5) {
+      ui.warning(
```
Comment (pabo99) [inline frontend/www/js/omegaup/common/index.ts:14]: I think the message types are inverted, this should be an error message:
```suggestion
      ui.error(
```

File: frontend/www/js/omegaup/common/index.ts [hunk 17]
```diff
@@ -2,9 +2,28 @@ import Homepage from '../components/homepage/Homepage.vue';
...
+        ui.formatString(T.accountVerifyWarning, {
+          days: commonPayload.pendingDaysForVerification,
+        }),
+      );
+    } else if (commonPayload.pendingDaysForVerification <= 7) {
+      ui.error(
```
Comment (pabo99) [inline frontend/www/js/omegaup/common/index.ts:20]: and it should be a warning message:
```suggestion
      ui.warning(
```

**Review by pabo99 — COMMENTED** [review]
It's almost done!

File: frontend/server/src/Controllers/Session.php [hunk 18]
```diff
@@ -327,6 +329,37 @@ private static function getCurrentSessionImpl(
...
+                ) && strtotime(
+                    $currentUser->birth_date
+                ) >= strtotime(
+                    '-13 year',
+                    $now
+                ) && !$currentUser->parent_verified
+            ) {
+                if (
+                    is_null($currentUser->parent_email_verification_deadline)
+                ) {
```
Comment (pabo99) [inline frontend/server/src/Controllers/Session.php:345]: Nit:
```suggestion
                if (is_null($currentUser->parent_email_verification_deadline)) {
```

File: frontend/server/src/Controllers/Session.php [hunk 19]
```diff
@@ -327,6 +329,37 @@ private static function getCurrentSessionImpl(
...
+                $pendingDaysForVerification = intval(
+                    floor(
+                        $diff / (60 * 60 * 24)
+                    )
+                );
+               if (
+                $pendingDaysForVerification < 0
+               ) {
+                throw new \OmegaUp\Exceptions\NotFoundException('U13HasNotBeenVerifiedForTheirParent');
+               }
```
Comment (pabo99) [inline frontend/server/src/Controllers/Session.php:361]:
```suggestion
               if ($pendingDaysForVerification < 0) {
                   throw new \OmegaUp\Exceptions\NotFoundException('U13HasNotBeenVerifiedForTheirParent');
               }
```

File: frontend/server/src/Controllers/Session.php [hunk 20]
```diff
@@ -327,6 +329,37 @@ private static function getCurrentSessionImpl(
...
+
```
Comment (pabo99) [inline frontend/server/src/Controllers/Session.php:362]:
```suggestion
```

File: frontend/server/src/Controllers/Session.php [hunk 21]
```diff
@@ -46,7 +46,7 @@ public static function getSessionManagerInstance(): \OmegaUp\SessionManager {
     public static function getFacebookLoginUrl(): string {
         $scopedFacebook = new ScopedFacebook();
         return $scopedFacebook->facebook->getAuthorizationUrl([
-            'scope' => ['email'],
+        'scope' => ['email'],
```
Comment (lhchavez) [inline frontend/server/src/Controllers/Session.php:49]:
```suggestion
            'scope' => ['email'],
```

same as in the other change: can the whitespace changes be minimized?

Comment (anmode) [reply frontend/server/src/Controllers/Session.php:49]: Sure! But i didn't touch this piece of code...
Should i didn't run this command ./stuff/lint.sh ?

Comment (lhchavez) [reply frontend/server/src/Controllers/Session.php:49]: your editor probably modified it by itself without you noticing and `./stuff/lint.sh` couldn't fix it.

Comment (anmode) [reply frontend/server/src/Controllers/Session.php:49]: Ohh okayy! But I had off the formatting of my editor. I noticed sometimes when i run this command these space problems occurred while sometimes not.

File: frontend/templates/en.lang [hunk 22]
```diff
@@ -1,9 +1,11 @@
+U13HasNotBeenVerifiedForTheirParent = "User U13 has not been verified by their parent So account will be blocked and unable to consume content"
```
Comment (lhchavez) [inline frontend/templates/en.lang:1]:
```suggestion
U13HasNotBeenVerifiedByTheirParent = "User under 13 has not been verified by their parent. This account will be blocked and unable to consume content"
```

File: frontend/templates/es.lang [hunk 23]
```diff
@@ -1,9 +1,11 @@
+U13HasNotBeenVerifiedForTheirParent = "El usuario U13 no ha sido verificado por su padre, por lo que la cuenta se bloqueará y no podrá consumir contenido"
```
Comment (lhchavez) [inline frontend/templates/es.lang:1]:
```suggestion
U13HasNotBeenVerifiedByTheirParent = "El usuario menor de 13 años no ha sido verificado por su figura parental, por lo que la cuenta se bloqueará y no podrá consumir contenido"
```

File: frontend/templates/es.lang [hunk 24]
```diff
@@ -1,9 +1,11 @@
...
 aboutToStart = "Estás a punto de iniciar el concurso. Una vez que hagas clic en entrar, tu tiempo iniciará."
 acceptPrivacyPolicy = "He leído y acepto la [Política de Uso y Privacidad](http://blog.omegaup.com/privacy-policy/) del sitio, así como el [Código de Conducta](https://blog.omegaup.com/codigo-de-conducta-en-omegaup/)."
 accountDeletionConfirmEmailBody = "El usuario %(username) confirmó eliminar su cuenta."
 accountDeletionConfirmEmailSubject = "Confirmación de eliminación de cuenta omegaUp."
 accountDeletionRequestEmailBody = "El usuario %(username) ha solicitado borrar su cuenta."
 accountDeletionRequestEmailSubject = "Solicitud de eliminación de cuenta omegaUp"
+accountVerifyWarning = "Su padre o tutor debe aprobar su cuenta, de lo contrario, se eliminará en %(days) días"
```
Comment (lhchavez) [inline frontend/templates/es.lang:8]:
```suggestion
accountVerifyWarning = "Su figura parental o tutor debe aprobar la cuenta, de lo contrario, se eliminará en %(days) días"
```

File: frontend/templates/en.lang (hunk 22 above)
Comment (lhchavez) [reply frontend/templates/en.lang:1]: @tvanessa @heduenas pueden checar estas cadenas?

File: frontend/server/src/Controllers/Session.php [hunk 25]
```diff
@@ -327,9 +329,38 @@ private static function getCurrentSessionImpl(
...
+                if ($pendingDaysForVerification < 0) {
+                    throw new \OmegaUp\Exceptions\NotFoundException(
+                        'U13HasNotBeenVerifiedForTheirParent'
+                    );
+                }
+            }
             $email = !is_null($currentUser->main_email_id) ?
-                \OmegaUp\DAO\Emails::getByPK($currentUser->main_email_id) :
-                null;
+            \OmegaUp\DAO\Emails::getByPK($currentUser->main_email_id) :
```
Comment (lhchavez) [inline frontend/server/src/Controllers/Session.php:362]: can these whitespace changes be reverted?

Comment (anmode) [reply frontend/server/src/Controllers/Session.php:362]: sure!

**Comment by anmode** [comment]
@lhchavez Did i have to make the test to get this rid out of codecov? Because that piece of code calculation is fine i hope! Pleae check it once

**Comment by lhchavez** [comment]
> @lhchavez Did i have to make the test to get this rid out of codecov? Because that piece of code calculation is fine i hope! Pleae check it once

yes, codecov flags code that is not executed inside a test. adding tests is the way to make those warnings go away.

File: frontend/server/src/Controllers/Session.php [hunk 26]
```diff
@@ -327,9 +329,38 @@ private static function getCurrentSessionImpl(
...
+                ) && !$currentUser->parent_verified
+            ) {
+                if (is_null($currentUser->parent_email_verification_deadline)) {
+                    throw new \OmegaUp\Exceptions\InvalidParameterException(
+                        'parameterNotFound',
+                        'parent_email_verification_deadline'
+                    );
+                }
+                $diff = $currentUser->parent_email_verification_deadline->time - $now;
+                $pendingDaysForVerification = intval(
```
Comment (lhchavez) [inline frontend/server/src/Controllers/Session.php:350]: let's add a test.

File: frontend/server/src/Controllers/Session.php [hunk 27]
```diff
@@ -327,9 +329,38 @@ private static function getCurrentSessionImpl(
...
+                if ($pendingDaysForVerification < 0) {
+                    throw new \OmegaUp\Exceptions\NotFoundException(
+                        'U13HasNotBeenVerifiedForTheirParent'
+                    );
+                }
+            }
             $email = !is_null($currentUser->main_email_id) ?
                 \OmegaUp\DAO\Emails::getByPK($currentUser->main_email_id) :
-                null;
+                 null;
```
Comment (lhchavez) [inline frontend/server/src/Controllers/Session.php:363]:
```suggestion
                null;
```

try to keep the unrelated whitespace changes to a minimum

File: frontend/server/src/Controllers/Session.php (hunk 26 above)
Comment (anmode) [reply frontend/server/src/Controllers/Session.php:350]: Sure! where I have to add test for it? and How can you make a test for this piece of code? I just want a hint, rest I'll make test.

File: frontend/server/src/Controllers/Session.php (hunk 27 above)
Comment (anmode) [reply frontend/server/src/Controllers/Session.php:363]: Sure!

**Comment by pabo99** [comment]
The PR #7358  attempts to solve the same issue. So, it can be closed in favor of the other one.
